# Ollama

[Import model](https://github.com/ollama/ollama/blob/main/docs/import.md)

[Modelfile](https://github.com/ollama/ollama/blob/main/docs/modelfile.md)

[API](https://github.com/ollama/ollama/blob/main/docs/api.md)

[API Python](https://github.com/ollama/ollama-python)

Ollama es un software basado en [llama.cpp](https://github.com/ggerganov/llama.cpp) para la inferencia de LLMs.

[llama.cpp](https://github.com/ggerganov/llama.cpp) es una librería para la inferencia de LLMs en C/C++, de modo que Ollama se ha construido por encima haciendo que el uso sea más sencillo para cualquier tipo de usuario. Es decir, es un wrapper para poder usar `llama.cpp` con una API de Python, JavaScript y terminal

## Instalación

Para instalar ollama en windows o mac tienes que descargarte un instalador desde su página de [descargas](https://ollama.com/download), sin embargo, si usas linux, como es mi caso, tienes que ejecutar el siguiente comando

``` bash
curl -fsSL https://ollama.com/install.sh | sh
```

## Run

Para correr un LLM basta con ejecutar `ollama run <LLM>`, por ejemplo puedo usar `llama3` mediante

In [1]:
!ollama run llama3

pulling manifest
pulling 6a0746a1ec1a... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏ 4.7 GB
pulling 4fa551d4f938... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  12 KB
pulling 8ab4849b038c... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  254 B
pulling 577073ffcc6c... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  110 B
pulling 3f8eb4da87fa... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  485 B
verifying sha256 digest
writing manifest
removing any unused layers
success
>>> hola
Hola! ¿Cómo estás?

>>> Send a message (/? for help)


Como se puede ver ollama funciona como docker, como no tenía descargado llama3, se lo ha descargado y luego podemos usarlo

## Modelos

Si queremos saber qué modelos hay disponibles podemos ir a su página de [modelos](https://ollama.com/library), donde se pueden ver todos los modelos disponibles

## Descargar un modelo

Si queremos descargarnos un modelo, lo haremos igual que con docker `ollama pull <LLM>`, por ejemplo, voy a descargarme `phi3`

In [2]:
!ollama pull phi3

pulling manifest
pulling 4fed7364ee3e... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏ 2.3 GB
pulling c608dc615584... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  149 B
pulling fa8235e5b48f... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏ 1.1 KB
pulling d47ab88b61ba... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  140 B
pulling f7eda1da5a81... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  485 B
verifying sha256 digest
writing manifest
removing any unused layers
success

## Ayuda

Si queremos ver todos los comandos que tenemos disponibles podemos usar `ollama -h`

In [3]:
!ollama -h

Large language model runner

Usage:
  ollama [flags]
  ollama [command]

Available Commands:
  serve       Start ollama
  create      Create a model from a Modelfile
  show        Show information for a model
  run         Run a model
  pull        Pull a model from a registry
  push        Push a model to a registry
  list        List models
  ps          List running models
  cp          Copy a model
  rm          Remove a model
  help        Help about any command

Flags:
  -h, --help      help for ollama
  -v, --version   Show version information

Use "ollama [command] --help" for more information about a command.


## Listar los modelos disponibles

Podemos ver en la ayuda que si hacemos `ollama list` podremos ver todos los modelos disponibles

In [4]:
!ollama list

NAME         	ID          	SIZE  	MODIFIED       
phi3:latest  	a2c89ceaed85	2.3 GB	4 minutes ago 	
llama3:latest	365c0bd3c000	4.7 GB	10 minutes ago	


Y como vemos solo tenemos los dos que hemos descargado hasta el momento

## Ver los modelos corriendo

Si hacemos `ollama ps` podemos ver los modelos que tenemos corriendo

In [6]:
!ollama ps

NAME         	ID          	SIZE  	PROCESSOR	UNTIL              
llama3:latest	365c0bd3c000	5.4 GB	100% GPU 	3 minutes from now	


Vemos que tenemos corriendo llama3 porque al principio hicimos `ollama run llama3`, pero si ahora lo intentamos con `phi3`

In [7]:
!ollama run phi3

>>> Send a message (/? for help)


Si ahora volvemos a hacer `ollama ps`

In [8]:
!ollama ps

NAME       	ID          	SIZE  	PROCESSOR	UNTIL              
phi3:latest	a2c89ceaed85	3.8 GB	100% GPU 	3 minutes from now	


Podemos ver que el modelo que está corriendo es `phi3`

## Modelos multimodales

En la lista de modelos de ollama hay modelos multimodales, como [moondream2](https://ollama.com/library/moondream), que siendo un modelo con solo 1.8B de parámetros tiene un desempeño muy bueno, así que vamos a descargarlo para probarlo

In [1]:
!ollama pull moondream

pulling manifest
pulling e554c6b9de01... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏ 828 MB
pulling 4cc1cb3660d8... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏ 909 MB
pulling c71d239df917... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  11 KB
pulling 4b021a3b4b4a... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏   77 B
pulling 9468773bdc1f... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏   65 B
pulling ba5fbb481ada... 100% ▕██████████████████████████████████████████████████████████████████████████████████████████▏  562 B
verifying sha256 digest
writing manifest
removing any unused layers
success

Una vez descargado vamos a preguntarle por esta imagen

![this_is_fine](https://maximofn.com/wp-content/uploads/2024/05/this_is_fine-scaled.webp)

Para ello tenemos tenemos que pasarle en el prompt la ruta donde tenemos la imagen

In [2]:
!ollama run moondream

>>> Puedes explicarme esta imagen? /home/wallabot/Documentos/web/portafolio/images/this_is_fine.png
Added image '/home/wallabot/Documentos/web/portafolio/images/this_is_fine.png'

The image is a comic strip featuring two dogs sitting at a table. One dog has its eyes closed, while the other one appears to be looking at something or someone in 
front of it. The scene takes place inside a house with a fireplace and a cup on the table. Above them, there's a speech bubble that says 'This is fine.'

## Prompts multilinea

Cuando necesitamos escribir más de una línea en nuestro prompt simplemente tenemos que meterlo entre comillas triples `"""`

In [1]:
!ollama run llama3

>>> """Hola, puedes hacerme un resumen de la siguiente conversación
...
... Alex: Bea, he estado revisando el modelo de clasificación de imágenes que entrenamos la semana pasada, y me parece que no está funcionando como esperábamos. Las métrica
... s de precisión y recall están por debajo de lo que habíamos estimado.
... 
... Bea: Sí, lo noté también. He estado analizando las predicciones erróneas y creo que el problema puede estar en el preprocesamiento de los datos. Algunos de los ejemplos
...  de entrenamiento parecen estar etiquetados incorrectamente.
... 
... Alex: Eso tiene sentido. ¿Crees que deberíamos aplicar alguna técnica de aumento de datos para balancear mejor las clases y mejorar la calidad de las etiquetas?
... 
... Bea: Podría funcionar. De hecho, estaba pensando en utilizar augmentations como rotaciones y traslaciones. También podríamos considerar el uso de GANs para generar más 
... ejemplos sintéticos, especialmente para las clases minoritarias.
... 
... Alex: Buena 

## Archivos GGUF

Como hemos dicho, ollama está basado en `llama.cpp`, por lo que el tipo de modelos que utiliza ollama son los que utiliza `llama.cpp` y son los archivos de tip `GGUF`.

Los archivos `GGUF` se refieren al formato de archivo `Generalized GPU-Friendly Unified Format`, que está diseñado para almacenar y 
ejecutar modelos de lenguaje grande (LLMs) de manera eficiente en hardware acelerado por GPU.

El formato GGUF es optimizado específicamente para almacenar y ejecutar modelos de lenguaje grande, como arquitecturas basadas en transformers. Proporciona una forma compacta y eficiente de representar los pesos del modelo, las activaciones y los grafos de computación, lo que 
permite tiempos de inferencia más rápidos y un uso reducido de memoria.

Los archivos GGUF suelen contener los siguientes componentes:

 1. Arquitectura del modelo: El archivo GGUF almacena la arquitectura del modelo, incluyendo el número de capas, tamaños ocultos y mecanismos de atención.
 2. Pesos: Los pesos del modelo se almacenan en un formato comprimido para reducir el uso de memoria. Es decir, los modelos están cuantizados
 3. Activaciones: El archivo GGUF contiene activaciones pre-computadas para ciertas capas o sub-capas, que pueden ser reutilizadas durante la inferencia.
 4. Grafo de computación: El grafo de computación representa la secuencia de operaciones necesarias para realizar la inferencia con el modelo.

Al utilizar archivos GGUF, los desarrolladores pueden implementar modelos de lenguaje grande en una variedad de hardware, 
incluyendo GPUs, TPUs y CPUs, mientras logran tiempos de inferencia más rápidos y un rendimiento mejorado.

Por ejemplo, el modelo llama3 de 70B de paŕametros en FP32 debería ocupar 280 GB, en FP16 debería ocupar 140 GB y en INT8 debería ocupar 70GB. Sin embargo, cuando nos lo vamos a descargar, en la web pone que ocupa 40 GB, por lo que al ocupar un poco más de la mitad podemos pensar que está cuantizado en INT4

Como tengo 2 GPUs de 24 GB lo podría usar entre las dos GPUs, si lo cargase en paralelo en las dos, por lo que me lo voy a descargar y lo voy a probar

In [1]:
!ollama pull llama3:70B

pulling manifest 
pulling 0bd51f8f0c97... 100% ▕████████████████████████████████████████████████████████████████████████████████████████████████▏  39 GB
pulling 4fa551d4f938... 100% ▕████████████████████████████████████████████████████████████████████████████████████████████████▏  12 KB
pulling 8ab4849b038c... 100% ▕████████████████████████████████████████████████████████████████████████████████████████████████▏  254 B
pulling 577073ffcc6c... 100% ▕████████████████████████████████████████████████████████████████████████████████████████████████▏  110 B
pulling ea8e06d28e47... 100% ▕████████████████████████████████████████████████████████████████████████████████████████████████▏  486 B
verifying sha256 digest 
writing manifest 
removing any unused layers 
success

Lo pruebo

In [2]:
!ollama run llama3:70B

>>> Hola, me puedes decir cómo pasar un array de numpy a un tensor de pytorch?
Sí!

En PyTorch, puedes convertir un arreglo de NumPy a un tensor utilizando la función `torch.from_numpy()`.

Aquí tienes un ejemplo:
```
import numpy as np
import torch

# Creamos un arreglo de NumPy
arr = np.array([1, 2, 3, 4, 5])

# Convertimos el arreglo a un tensor de PyTorch
tensor = torch.from_numpy(arr)

print(tensor)  # Output: tensor([1, 2, 3, 4, 5])
```
Ten en cuenta que la función `torch.from_numpy()` devuelve un tensor que comparte la memoria con el arreglo de NumPy original. Esto significa que si 
modificas el tensor, también estarás modificando el arreglo de NumPy original.

Si deseas crear un tensor independiente del arreglo de NumPy, puedes utilizar la función `torch.tensor()` y pasarle el arreglo de NumPy como parámetro:
```
tensor = torch.tensor(arr)
```
De esta manera, el tensor creado será una copia independiente del arreglo de NumPy original.

Espero que esto te ayude!

No solo contesta bien, sino que va rapidísimo

Además se comprueba que ha cargado el modelo en mis dos GPUs en paralelo

Por un lado se puede suponer, porque en una sola no entraría como he comentado antes. Pero es que además he creado un [monitor de GPUs](https://github.com/maximofn/gpu_monitor), con el que puedo ver que están las dos GPUs casi llenas enteras

De aquí podemos sacar otra conclusión, y es que mediante los archivos `GGUF` podemos ejecutar los LLMs de la manera más eficiente en nuestros ordenadores, en mis caso ha cargado en paralelo el LLM en mis dos GPUs

## Modelfile

Ahora que hemos visto lo que son los archivos `GGUF` podemos ver cómo crear nuevos a partir de otros. Al igual que con docker podemor crear imagenes a partir de otras con un dockerfile, ollama nos permite crear modelos a partir de otros con un modelfile

Para esto vamos a partir del archivo `Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf` que es el archivo `GGUF` del modelo `NousResearch/Hermes-2-Theta-Llama-3-8B-GGUF`. Este modelo es una versión de llama3 de 8B de parámetros

A la hora de crear modelos y subirlos a internet lo ideal es ponerle nombres descriptivos, por ejemplo, en este caso vemos que tenemos en el nombre

 * Llama-3-Instruct: Lo que quiere decir que es un modelo que parte de llama3 8B Instruct
 * Merged: Es un modelo mergeado
 * DPO: Se ha hecho el alineamiento con DPO
 * Q4: Está cuantizado a 4 bits

Primero vamos a crear la carpeta donde vamos a crear el modelfile

In [3]:
!mkdir ollama_modelfile

Ahora creamos el modelfile

In [10]:
!touch ollama_modelfile/Modelfile

Y ahora nos descargamos el archivo `GGUF`

In [9]:
!wget https://huggingface.co/NousResearch/Hermes-2-Theta-Llama-3-8B-GGUF/resolve/main/Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf -O ollama_modelfile/Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf

--2024-05-22 19:16:24--  https://huggingface.co/NousResearch/Hermes-2-Theta-Llama-3-8B-GGUF/resolve/main/Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf
Resolviendo huggingface.co (huggingface.co)... 54.192.95.21, 54.192.95.79, 54.192.95.70, ...
Conectando con huggingface.co (huggingface.co)[54.192.95.21]:443... conectado.
Petición HTTP enviada, esperando respuesta... 302 Found
Ubicación: https://cdn-lfs-us-1.huggingface.co/repos/48/0d/480d1b9ea3df81b6cf974d62b6bc04f402ac5600a1dfdde7bbb9e819a111c572/762b9371a296ab2628592b9462dc676b27d881a3402816492801641a437669b3?response-content-disposition=attachment%3B+filename*%3DUTF-8%27%27Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf%3B+filename%3D%22Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf%22%3B&Expires=1716657384&Policy=eyJTdGF0ZW1lbnQiOlt7IkNvbmRpdGlvbiI6eyJEYXRlTGVzc1RoYW4iOnsiQVdTOkVwb2NoVGltZSI6MTcxNjY1NzM4NH19LCJSZXNvdXJjZSI6Imh0dHBzOi8vY2RuLWxmcy11cy0xLmh1Z2dpbmdmYWNlLmNvL3JlcG9zLzQ4LzBkLzQ4MGQxYjllYTNkZjgxYjZjZjk3

Una vez descargado empezamos a crear el modelfile

### FROM

Primero tenemos que especificar desde qué modelo partimos

Podemos partir desde un modelo que esté en el hub de ollama, por ejemplo podríamos hacer

``` yaml
FROM llama3
```

Pero como vamos a partir desde el archivo que nos hemos descargado, lo especificamos mediante el path

``` yaml
FROM ./Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf
```

In [11]:
%%writefile ollama_modelfile/Modelfile
FROM ./Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf

Overwriting ollama_modelfile/Modelfile


### PARAMETER

Ahora tenemos que establecer los valores de los parámetros del LLM, por ejemplo, si queremos modificar el valor de la temperatura ponemos

``` yaml
PARAMETER temperature 0.5
```

Los parámetros que podemos modificar son

| Parameter | Description | Value Type | Example Usage |
| --- | --- | --- | --- |
| mirostat | Enable Mirostat sampling for controlling perplexity. (default: 0, 0 = disabled, 1 = Mirostat, 2 = Mirostat 2.0) | int | mirostat 0 |
| mirostat_eta | Influences how quickly the algorithm responds to feedback from the generated text. A lower learning rate will result in slower adjustments, while a higher learning rate will make the algorithm more responsive. (Default: 0.1) | float | mirostat_eta 0.1 |
| mirostat_tau | Controls the balance between coherence and diversity of the output. A lower value will result in more focused and coherent text. (Default: 5.0) | float | mirostat_tau 5.0 |
| num_ctx | Sets the size of the context window used to generate the next token. (Default: 2048) | int | num_ctx 4096 |
| repeat_last_n | Sets how far back for the model to look back to prevent repetition. (Default: 64, 0 = disabled, -1 = num_ctx) | int | repeat_last_n 64 |
| repeat_penalty | Sets how strongly to penalize repetitions. A higher value (e.g., 1.5) will penalize repetitions more strongly, while a lower value (e.g., 0.9) will be more lenient. (Default: 1.1) | float | repeat_penalty 1.1 |
| temperature | The temperature of the model. Increasing the temperature will make the model answer more creatively. (Default: 0.8) | float | temperature 0.7 |
| seed | Sets the random number seed to use for generation. Setting this to a specific number will make the model generate the same text for the same prompt. (Default: 0) | int | seed 42 |
| stop | Sets the stop sequences to use. When this pattern is encountered the LLM will stop generating text and return. Multiple stop patterns may be set by specifying multiple separate stop parameters in a modelfile. | string | stop "AI assistant:" |
| tfs_z | Tail free sampling is used to reduce the impact of less probable tokens from the output. A higher value (e.g., 2.0) will reduce the impact more, while a value of 1.0 disables this setting. (default: 1) | float | tfs_z 1 |
| num_predict | Maximum number of tokens to predict when generating text. (Default: 128, -1 = infinite generation, -2 = fill context) | int | num_predict 42 |
| top_k | Reduces the probability of generating nonsense. A higher value (e.g. 100) will give more diverse answers, while a lower value (e.g. 10) will be more conservative. (Default: 40) | int | top_k 40 |
| top_p | Works together with top-k. A higher value (e.g., 0.95) will lead to more diverse text, while a lower value (e.g., 0.5) will generate more focused and conservative text. (Default: 0.9) | float | top_p 0.9 |

Vamos a hacer al modelo más creativo subiéndole la temperatura

In [13]:
%%writefile ollama_modelfile/Modelfile
FROM ./Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf

PARAMETER temperature 1

Overwriting ollama_modelfile/Modelfile


### SYSTEM

Podemos condicionar al modelo con un mensaje de sistema

In [1]:
%%writefile ollama_modelfile/Modelfile
FROM ./Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf

PARAMETER temperature 1

SYSTEM """Eres un asistente de IA, que quiere ayudar, pero no tiene muchas habilidades sociales"""

Overwriting ollama_modelfile/Modelfile


### LICENSE

Sirve para establecer la licencia del modelo

In [15]:
%%writefile ollama_modelfile/Modelfile
FROM ./Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf

PARAMETER temperature 1

SYSTEM """Eres un asistente de IA, que quiere ayudar, pero no tiene muchas habilidades sociales"""

LICENSE """Apache-2.0"""

Overwriting ollama_modelfile/Modelfile


### MESSAGE

Podemos pasarle un historial de chat

In [2]:
%%writefile ollama_modelfile/Modelfile
FROM ./Hermes-2-Pro-Llama-3-Instruct-Merged-DPO-Q4_K_M.gguf

PARAMETER temperature 1

SYSTEM """Eres un asistente de IA, que quiere ayudar, pero no tiene muchas habilidades sociales"""

LICENSE """Apache-2.0"""

MESSAGE user Sabes cómo pasar un array de numpy a un tensor de torch?
MESSAGE assistant Sí
MESSAGE user Y me lo podrías decir?
MESSAGE assistant Sí
MESSAGE user Dímelo
MESSAGE assistant Te lo puedo decir

Overwriting ollama_modelfile/Modelfile


### Crear el modelo

Una vez escrito el modelfile podemos crear el modelo mediante

``` bash
ollama create <model-name> -f <path-of-the-modelfile>'
```

In [3]:
!ollama create assistanmal -f ollama_modelfile/Modelfile

transferring model data
using existing layer sha256:762b9371a296ab2628592b9462dc676b27d881a3402816492801641a437669b3
creating new layer sha256:ba85fc82591e657e639f779f4fdc5e6e7cb977f9d36487d67cefe589a8af670e
creating new layer sha256:2af71558e438db0b73a20beab92dc278a94e1bbe974c00c1a33e3ab62d53a608
creating new layer sha256:20dc17cb752058f95ad600722883b22f3328ca8ccccbb667d5e07991ae1985c7
using existing layer sha256:d8ba2f9a17b3bbdeb5690efaa409b3fcb0b56296a777c7a69c78aa33bbddf182
creating new layer sha256:2156b3b25c67a5712e4c11c889ef91c1625c4ea7b59c5c81ab2a7aa451535acf
writing manifest
success

Vemos que tenemos una salida muy similar a la que obtenemos cuando creamos una imagen de docker

Si ahora listamos los modelos que tenemos disponibles

In [4]:
!ollama list

NAME              	ID          	SIZE  	MODIFIED           
assistanmal:latest	84222b0e11b8	4.9 GB	About a minute ago	
llama3:70b        	786f3184aec0	39 GB 	12 hours ago      	
moondream:latest  	55fc3abd3867	1.7 GB	14 hours ago      	
phi3:latest       	a2c89ceaed85	2.3 GB	40 hours ago      	
llama3:latest     	365c0bd3c000	4.7 GB	40 hours ago      	


Vemos que está el asistente que acabamos de crear. Vamos a probarlo

In [5]:
!_ollama run assistantmal

>>> Sabes cómo pasar un array de numpy a un tensor de torch?
Sí

>>> Y me lo podrías decir?
Sí

>>> Dímelo
Te lo puedo decir

>>> Me puedes responder?
¡Claro!

Para pasar un arreglo (array) de NumPy a un tensor de PyTorch, simplemente necesitas importar la librería `torch` y usar la función `tensor()` para convertir el 
arreglo a un tensor. Aquí tienes un ejemplo:

```python
import torch

# Importar arreglo de numpy
import numpy as np

# Crear arreglo de numpy
numpy_array = np.array([1, 2, 3])

# Convertir arreglo de numpy a tensor de PyTorch
pytorch_tensor = torch.tensor(numpy_array)

print(pytorch_tensor)
```

En este ejemplo, creamos un arreglo de numpy con los valores `[1, 2, 3]`. Luego, lo convertimos a un tensor de PyTorch usando la función `tensor()`. Finalmente, 
imprimimos el tensor resultante.

Ten en cuenta...

Lo he tenido que parar porque no para de responder. No estaba buscando hacer el mejor assistente